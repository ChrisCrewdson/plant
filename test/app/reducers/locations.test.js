// @ts-ignore - static hasn't been defined on seamless types yet.
const seamless = require('seamless-immutable').static;
const locations = require('../../../app/reducers/locations');
const { actionFunc } = require('../../../app/actions');

describe('/app/reducers/locations', () => {
  const stateA = seamless.from({
    1: {
      _id: '1',
      plantIds: ['one'],
    },
    2: {
      _id: '2',
      name: 'xxx',
      plantIds: ['xxx'],
    },
    3: {
      _id: '3',
    },
  });

  test('should load locations', () => {
    const payload = {
      2: {
        _id: '2',
        name: 'twenty-two',
        plantIds: ['two', 'twenty'],
      },
      3: {
        _id: '3',
        name: 'thirty-three',
        plantIds: ['three', 'thirty'],
      },
      4: {
        _id: '4',
        name: 'four',
      },
    };

    const actual = locations(stateA, actionFunc.loadLocationsSuccess(payload));
    expect(actual).toMatchSnapshot();
    expect(seamless.isImmutable(actual)).toBe(true);
  });

  test('should load locations with undefined payload', () => {
    const actual = locations(stateA, actionFunc.loadLocationsSuccess(undefined));
    expect(actual).toMatchSnapshot();
    // The same object should be returned
    expect(actual).toBe(stateA);
    expect(seamless.isImmutable(actual)).toBe(true);
  });

  test('should handle a createPlantRequest for an existing plant at location', () => {
    const plant = {
      locationId: '1',
      _id: 'plant-1',
    };
    const actual = locations(stateA, actionFunc.createPlantRequest(plant));
    expect(actual).toMatchSnapshot();
    expect(actual).not.toBe(stateA);
    expect(seamless.isImmutable(actual)).toBe(true);
  });

  test('should handle a createPlantRequest for a missing plant at location', () => {
    const plant = {
      locationId: '7',
      _id: 'plant-1',
    };
    const actual = locations(stateA, actionFunc.createPlantRequest(plant));
    expect(actual).toMatchSnapshot();
    expect(actual).toBe(stateA);
    expect(seamless.isImmutable(actual)).toBe(true);
  });

  test('should handle a createPlantRequest when location does not have plantIds', () => {
    const plant = {
      locationId: '3',
      _id: 'plant-1',
    };
    const actual = locations(stateA, actionFunc.createPlantRequest(plant));
    expect(actual).toMatchSnapshot();
    expect(actual).not.toBe(stateA);
    expect(seamless.isImmutable(actual)).toBe(true);
  });

  test('should handle a loadPlantsSuccess for an array of plants', () => {
    const plants = [{
      locationId: '1',
      _id: 'plant-1',
    }, { // include a duplicate to make sure it doesn't end up in state
      locationId: '1',
      _id: 'plant-1',
    }, {
      locationId: '7',
      _id: 'plant-7',
    }, {
      locationId: '1',
      _id: 'one', // a duplicate from stateA
    }];
    const actual = locations(stateA, actionFunc.loadPlantsSuccess(plants));
    expect(actual).toMatchSnapshot();
    expect(seamless.isImmutable(actual)).toBe(true);
  });

  test('should handle a loadPlantsSuccess for an empty array of plants', () => {
    /** @type {UiPlantsValue[]} */
    const plants = [];
    const actual = locations(stateA, actionFunc.loadPlantsSuccess(plants));
    expect(actual).toMatchSnapshot();
    expect(actual).toBe(stateA);
    expect(seamless.isImmutable(actual)).toBe(true);
  });

  test('should handle a deletePlantRequest', () => {
    const deletePlant = {
      locationId: '1',
      plantId: 'one',
    };
    const actual = locations(stateA, actionFunc.deletePlantRequest(deletePlant));
    expect(actual).toMatchSnapshot();
    expect(seamless.isImmutable(actual)).toBe(true);
  });

  test('should handle a deletePlantRequest when locationId is missing', () => {
    const deletePlant = {
      locationId: '7',
      plantId: 'one',
    };
    const actual = locations(stateA, actionFunc.deletePlantRequest(deletePlant));
    expect(actual).toMatchSnapshot();
    expect(actual).toBe(stateA);
    expect(seamless.isImmutable(actual)).toBe(true);
  });
});
