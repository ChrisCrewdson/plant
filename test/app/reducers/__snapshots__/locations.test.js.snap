// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`/app/reducers/locations should handle a createPlantRequest for a missing plant at location 1`] = `
Object {
  "1": Object {
    "_id": "1",
    "plantIds": Array [
      "one",
    ],
  },
  "2": Object {
    "_id": "2",
    "name": "xxx",
    "plantIds": Array [
      "xxx",
    ],
  },
  "3": Object {
    "_id": "3",
  },
}
`;

exports[`/app/reducers/locations should handle a createPlantRequest for an existing plant at location 1`] = `
Object {
  "1": Object {
    "_id": "1",
    "plantIds": Array [
      "one",
      "plant-1",
    ],
  },
  "2": Object {
    "_id": "2",
    "name": "xxx",
    "plantIds": Array [
      "xxx",
    ],
  },
  "3": Object {
    "_id": "3",
  },
}
`;

exports[`/app/reducers/locations should handle a createPlantRequest when location does not have plantIds 1`] = `
Object {
  "1": Object {
    "_id": "1",
    "plantIds": Array [
      "one",
    ],
  },
  "2": Object {
    "_id": "2",
    "name": "xxx",
    "plantIds": Array [
      "xxx",
    ],
  },
  "3": Object {
    "_id": "3",
    "plantIds": Array [
      "plant-1",
    ],
  },
}
`;

exports[`/app/reducers/locations should handle a deletePlantRequest 1`] = `
Object {
  "1": Object {
    "_id": "1",
    "plantIds": Array [],
  },
  "2": Object {
    "_id": "2",
    "name": "xxx",
    "plantIds": Array [
      "xxx",
    ],
  },
  "3": Object {
    "_id": "3",
  },
}
`;

exports[`/app/reducers/locations should handle a deletePlantRequest when locationId is missing 1`] = `
Object {
  "1": Object {
    "_id": "1",
    "plantIds": Array [
      "one",
    ],
  },
  "2": Object {
    "_id": "2",
    "name": "xxx",
    "plantIds": Array [
      "xxx",
    ],
  },
  "3": Object {
    "_id": "3",
  },
}
`;

exports[`/app/reducers/locations should handle a loadPlantsSuccess for an array of plants 1`] = `
Object {
  "1": Object {
    "_id": "1",
    "plantIds": Array [
      "one",
      "plant-1",
    ],
  },
  "2": Object {
    "_id": "2",
    "name": "xxx",
    "plantIds": Array [
      "xxx",
    ],
  },
  "3": Object {
    "_id": "3",
  },
}
`;

exports[`/app/reducers/locations should handle a loadPlantsSuccess for an empty array of plants 1`] = `
Object {
  "1": Object {
    "_id": "1",
    "plantIds": Array [
      "one",
    ],
  },
  "2": Object {
    "_id": "2",
    "name": "xxx",
    "plantIds": Array [
      "xxx",
    ],
  },
  "3": Object {
    "_id": "3",
  },
}
`;

exports[`/app/reducers/locations should load locations 1`] = `
Object {
  "1": Object {
    "_id": "1",
    "plantIds": Array [
      "one",
    ],
  },
  "2": Object {
    "_id": "2",
    "name": "twenty-two",
    "plantIds": Array [
      "two",
      "twenty",
    ],
  },
  "3": Object {
    "_id": "3",
    "name": "thirty-three",
    "plantIds": Array [
      "three",
      "thirty",
    ],
  },
  "4": Object {
    "_id": "4",
    "name": "four",
    "plantIds": Array [],
  },
}
`;

exports[`/app/reducers/locations should load locations with undefined payload 1`] = `
Object {
  "1": Object {
    "_id": "1",
    "plantIds": Array [
      "one",
    ],
  },
  "2": Object {
    "_id": "2",
    "name": "xxx",
    "plantIds": Array [
      "xxx",
    ],
  },
  "3": Object {
    "_id": "3",
  },
}
`;
